presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -458,10 +517,244 @@ else if (!result.getBlocked().isDone()) {
         return stream(forTree(StreamingPlanSection::getChildren).depthFirstPreOrder(sectionedPlan))
                 // get all sections ready for execution
                 .filter(this::isReadyForExecution)
+                /* .map(this::fetchTemporaryTableStatistics) */
+                .map(this::tryCostBasedOptimize)
                 .limit(maxConcurrentMaterializations - runningPlanSections)
                 .collect(toImmutableList());
     }
 
+    /**
+     * Utility function to invoke runtime cost-based optimizer, which determines if the probe and build side of a JoinNode should be swapped,
+     * based on the statistics of the temporary table holding materialized exchange outputs from finished children sections.
+     */
+    private StreamingPlanSection tryCostBasedOptimize(StreamingPlanSection section)
+    {
+        // no need to do runtime optimization if no materialized exchange data is utilized by the section.
+        if (section.getChildren().isEmpty()) {
+            return section;
+        }
+
+        Map<PlanFragment, PlanFragment> oldToNewFragment = stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan()))
+                // filter leaf stages
+                .filter(plan -> plan.getChildren().isEmpty())
```

Comment:
ideally all this logic to filter relevant plans should go into the optimizer rule

Reply:
Yeah, I thought about moving all the filtering logic inside the rule, but my concern is it seems not possible to easily tell from the iterative optimizer output if the returned plan is actually changed or not. If we are not able to tell, then for each ready section, we will waste time rebuilding the whole section and stageExecution/scheduler etc. even the optimizer rule is not fired at all.